File: logic/src/main/java/com/tp/holdem/logic/HandRankComparator.java
package com.tp.holdem.logic;

import com.tp.holdem.logic.model.Card;
import com.tp.holdem.logic.model.HandRank;
import com.tp.holdem.logic.model.Hands;
import io.vavr.collection.List;

import java.util.Comparator;

class HandRankComparator implements Comparator<HandRank> {
	static final HandRankComparator INSTANCE = new HandRankComparator();

	private HandRankComparator() {
	}

	@Override
	public int compare(final HandRank one, final HandRank two) {
		final int compareHands = Comparator.comparing(Hands::value).compare(one.getHand(), two.getHand());

		if (compareHands != 0)
			return compareHands;

		final Hands firstHand = one.getHand();
		final List<Card> firstCards = List.ofAll(one.getCardsThatMakeHand());
		final Hands secondHand = two.getHand();
		final List<Card> secondCards = List.ofAll(two.getCardsThatMakeHand());

		if (firstHand == Hands.HIGH_CARD)
			return HandComparators.highestKicker().compare(firstCards, secondCards);
		if (firstHand == Hands.PAIR)
			return HandComparators.highestPair().compare(firstCards, secondCards);
		if (firstHand == Hands.TWO_PAIR)
			return HandComparators.highestPairs().compare(firstCards, secondCards);
		if (firstHand == Hands.THREE_OF_A_KIND)
			return HandComparators.highestThree().compare(firstCards, secondCards);
		if (firstHand == Hands.STRAIGHT)
			return HandComparators.highestStraightKicker().compare(firstCards, secondCards);
		if (firstHand == Hands.FLUSH)
			return HandComparators.highestKicker().compare(firstCards, secondCards);
		if (firstHand == Hands.FULL_HOUSE)
			return HandComparators.highestFullHouse().compare(firstCards, secondCards);
		if (firstHand == Hands.FOUR_OF_A_KIND)
			return HandComparators.highestFour().compare(firstCards, secondCards);
		if (firstHand == Hands.STRAIGHT_FLUSH)
			return HandComparators.highestKicker().compare(firstCards, secondCards);
		if (firstHand == Hands.ROYAL_FLUSH)
			return 0;

		return 0;
	}
}

File: logic/src/main/kotlin/com/tp/holdem/logic/hands/HandRankComparator.kt
package com.tp.holdem.logic.hands

import com.tp.holdem.model.HandRank
import com.tp.holdem.model.Hands
import java.util.*

internal object HandRankComparator : Comparator<HandRank> {
    override fun compare(one: HandRank, two: HandRank): Int {
        val compareHands = compareValuesBy(one.hand, two.hand, Hands::value)

        if (compareHands != 0)
            return compareHands

        val comparator = when (one.hand) {
            Hands.HIGH_CARD -> HandComparators.highestKicker
            Hands.PAIR -> HandComparators.highestPair
            Hands.TWO_PAIR -> HandComparators.highestPairs
            Hands.THREE_OF_A_KIND -> HandComparators.highestThree
            Hands.STRAIGHT -> HandComparators.highestStraightKicker
            Hands.FLUSH -> HandComparators.highestKicker
            Hands.FULL_HOUSE -> HandComparators.highestFullHouse
            Hands.FOUR_OF_A_KIND -> HandComparators.highestFour
            Hands.STRAIGHT_FLUSH -> HandComparators.highestKicker
            Hands.ROYAL_FLUSH -> HandComparators.highestKicker
        }

        return compareValuesBy(one, two, comparator, HandRank::cardsThatMakeHand)
    }
}
